app/gui2/src/components/ComponentBrowser/input.ts

```diff
@@ -269,27 +316,62 @@ export function useComponentBrowserInput(graphDb: GraphDb = useGraphStore().db)
     return oprAppSpacing + entry.name
   }
 
-  /** All changes to the qualified name already written by the user.
-   *
-   * See `inputChangesAfterApplying`. */
-  function* qnChangesAfterApplying(
-    entry: SuggestionEntry,
-  ): Generator<{ range: [number, number]; str: string }> {
-    if (entry.selfType != null) return []
-    if (entry.kind === SuggestionKind.Local || entry.kind === SuggestionKind.Function) return []
-    if (context.value.type === 'changeLiteral') return []
-    if (context.value.oprApp == null) return []
-    const writtenQn = qnIdentifiers(context.value.oprApp).reverse()
-
-    let containingQn =
-      entry.kind === SuggestionKind.Module
-        ? qnParent(entry.definedIn)
-        : entry.memberOf ?? entry.definedIn
-    for (const ident of writtenQn) {
-      if (containingQn == null) break
-      const [parent, segment] = qnSplit(containingQn)
-      yield { range: parsedTreeRange(ident), str: segment }
-      containingQn = parent
+  /** All changes to the qualified name already written by the user. */
+  function qnChanges(entry: SuggestionEntry): {
+    changes: Change[]
+    requiredImport: QualifiedName | null
+  } {
+    if (entry.selfType != null) return { changes: [], requiredImport: null }
+    if (entry.kind === SuggestionKind.Local || entry.kind === SuggestionKind.Function)
+      return { changes: [], requiredImport: null }
+    if (context.value.type !== 'changeLiteral' && context.value.oprApp != null) {
+      // 1. Find the index of last identifier from the original qualified name that is already written by the user.
+      const qn = entryQn(entry)
+      const identifiers = qnIdentifiers(context.value.oprApp)
+      const writtenSegments = identifiers.map((ident) => ident.repr())
+      const allSegments = qnSegments(qn)
+      const windowSize = writtenSegments.length
+      let indexOfAlreadyWrittenSegment = undefined
+      for (let right = allSegments.length; right >= windowSize; right--) {
+        const left = right - windowSize
+        if (equalFlat(allSegments.slice(left, right), writtenSegments)) {
+          indexOfAlreadyWrittenSegment = left
+          break
+        }
+      }
+      if (indexOfAlreadyWrittenSegment == null) {
+        // We didn’t find the exact match, which probably means the user has written
+        // partial names, like `Dat.V.` instead of `Data.Vector`.
```

Comment:
Is there a test case covering this or similar scenario? I can't find one.

Reply:
Got a response via discord: It is part of already existing test suite, therefore it is not immediately visible in the diff here.